inference/primitive.go

```diff
@@ -38,67 +42,300 @@ type primitiveFullTrigger struct {
+// encodes the package and site representations, and more importantly, the position information to
+// uniquely identify the site. However, in incremental build systems, when analyzing the downstream
+// package, the `upstream/main.go` file in the `analysis.Pass.Fset` will not have complete line and
+// column information. Instead, the [archive importer] injects 65535 fake "lines" into the file,
+// and the object we get for `upstream.GlobalVar` will contain completely different position
+// information due to this hack (in practice, we have observed that the lines for the objects are
+// correct, but not others). This leads to mismatches in our inference engine: we cannot find
+// nilabilities of upstream sites in the imported InferredMap since the positions of the primitive
+// sites are different. The primitivizer here contains logic to fix such discrepancies.
+// Note that this fix (mapping) is two-ways:
+// (1) Local incorrect upstream position to correct one for matching upstream sites in the
+// inferred map (such that all position information in primitiveSite is always correct). This is
+// done via primitivizer.site.
+// (2) Correct upstream position to local incorrect one for error reporting purposes (the analysis
+// framework only allows reporting via local token.Pos). This is done via primitivizer.sitePos.
+//
+// [archive importer]: https://github.com/golang/tools/blob/fa12f34b4218307705bf0365ab7df7c119b3653a/internal/gcimporter/bimport.go#L59-L69
+type primitivizer struct {
+	pass *analysis.Pass
+	// upstreamObjPositions maps "<pkg repr>.<object path>" to the correct position.
+	upstreamObjPositions map[string]token.Position
+	// files maps the file name (modulo the possible build-system prefix) to the token.File object
+	// for faster lookup when converting correct upstream position back to local token.Pos for
+	// reporting purposes.
+	files map[string]fileInfo
+	// curDir is the current working directory, which is used to trim the prefix (e.g., bazel
+	// random sandbox prefix) from the file names for cross-package references.
+	curDir string
+}
+
+// newPrimitivizer returns a new and properly-initialized primitivizer.
+func newPrimitivizer(pass *analysis.Pass) *primitivizer {
+	// To tackle the position discrepancies for upstream sites, we have added an ObjectPath field
+	// to primitiveSite, which can be used to uniquely identify an exported object relative to the
+	// package. Then, we simply cache the correct position information when importing
+	// InferredMaps, since the positions collected in the upstream analysis are always correct.
+	// Later when querying upstream objects in downstream analysis, we can look up the cache and
+	// fill in the correct position in the returned primitive site instead.
+
+	// Create a cache for upstream object positions.
+	upstreamObjPositions := make(map[string]token.Position)
+	for _, pkgFact := range pass.AllPackageFacts() {
+		importedMap, ok := pkgFact.Fact.(*InferredMap)
```

Comment:
We are now scanning for all `InferredMap` facts exported by NilAway in upstream dependencies multiple times, right? Is this fast enough or should we be doing this wherever else we load these maps?

Reply:
Does the comment still hold? If yes, did the performance evaluation succeed internally?

Reply:
Yes, our internal performance evaluation shows the build time overhead increases from 2% to ~2.9%. However, this increase is mostly due to more sites being correctly connected and more triggers being generated, instead of the extra map access here 😃